src/battle-log.ts

```diff
@@ -93,16 +93,54 @@ class BattleLog {
 			}
 			break;
 
-		case 'join': case 'j': {
+		case 'join': case 'j': case 'leave': case 'l': {
 			const user = BattleTextParser.parseNameParts(args[1]);
-			divHTML = '<small>' + BattleLog.escapeHTML(user.group + user.name) + ' joined.</small>';
-			break;
-		}
-		case 'leave': case 'l': {
-			const user = BattleTextParser.parseNameParts(args[1]);
-			divHTML = '<small>' + BattleLog.escapeHTML(user.group + user.name) + ' left.</small>';
+			const formattedUser = BattleLog.escapeHTML(user.group + user.name);
+			const allMessages = (preempt ? this.preemptElem : this.innerElem).children;
+			const joinMessage = allMessages.item(allMessages.length - 1);
+			const rawJoinsLeaves =  joinMessage?.innerHTML.split(/<\/?small>/).join('');
+			const isJoin = args[0].includes('j');
+			let joins: string[] = [];
+			let leaves: string[] = [];
+			let rawLeaves;
+
+			if (rawJoinsLeaves?.includes(' joined')) {
+				const joinSplit = rawJoinsLeaves?.split(';')[0].replace(' joined', '').split(' and ');
+				joins = joinSplit[0].split(', ');
+				if (joinSplit.length > 1) joins.push(joinSplit[1]);
+				if (rawJoinsLeaves.includes('; ')) rawLeaves = rawJoinsLeaves?.split('; ')[1];
+			} else {
+				if (rawJoinsLeaves) rawLeaves = rawJoinsLeaves;
+			}
+
+			if (rawLeaves) {
+				const leaveSplit = rawLeaves.replace(' left', '').split(' and ');
```

Comment:
Okay I definitely don't think the correct way to do this is to text-parse the previous line.

I would recommend a `BattleLog#joinLeave` property; like how `client-check` does it.

Reply:
You can basically just copy/paste the implementation from `client-chat` but rewrite the jQuery to be raw JavaScript.

Reply:
Here's how I'd do it:

```ts
joinLeave: {
  joins: string[],
  leaves: string[],
  element: HTMLDivElement,
} | null;
```

If `this.joinLeave` is `null`, you add a new element and populate it. Otherwise, you set the `element.innerText` to the new message that you generate from `joinLeave.joins` and `joinLeave.leaves`. You set `this.joinLeave = null` whenever you add a non-join/leave message.

Reply:
Sure.